Add tests for knocking

anoadragon453 opened the pull request:
This PR adds tests for knocking. They are currently hidden behind a build tag. To run the tests, you'll need to add `-tags msc2403` to your command line args to include these tests. The build flag, as well as all of the `xyz.amorgan.knock` references should be remove/modified after MSC2403 makes it into a spec version.

* MSC: https://github.com/matrix-org/matrix-doc/pull/2403
* Synapse implementation: https://github.com/matrix-org/synapse/pull/6739

Kegsay commented on internal/client/client.go:
```diff
@@ -101,6 +101,14 @@ func (c *CSAPI) SyncUntilTimelineHas(t *testing.T, roomID string, check func(gjs
+// SyncUntilKnockHas blocks and continually calls /sync until the `check` function returns true.
+// If the `check` function fails the test, the failing event will be automatically logged.
+// Will time out after CSAPI.SyncUntilTimeout.
+func (c *CSAPI) SyncUntilKnockHas(t *testing.T, roomID string, check func(gjson.Result) bool) {
```
I don't think knocking is _that_ common that it should pollute the `CSAPI` client. Move this to the test file please.

Kegsay replied:
If you require `syncUntil` logic, make that public instead.

Kegsay commented on tests/knocking_test.go:
```diff
@@ -0,0 +1,194 @@
+// +build knock
```
Prefer the MSC number please, as feature names are ambiguous (particularly if there are multiple competing MSCs for the same feature.

Kegsay replied:
Also please add the MSC number to the filename, e.g like https://github.com/matrix-org/complement/blob/kegan/msc2836/tests/msc2836_test.go

anoadragon453 replied:
Good point, much more scalable indeed.

clokep commented on tests/msc2403_test.go:
```diff
@@ -0,0 +1,241 @@
+	})
+
+	if !federation {
+		// Rescinding a knock over federation is currently not supported in Synapse
```
Isn't it not even defined in the spec?

anoadragon453 replied:
*Rejecting* a knock over federation is currently not possible, but rescinding a knock over federation can work (implementations just need to figure out which homeserver to send the leave event through).

https://github.com/matrix-org/matrix-doc/pull/2403/files#diff-9dcacf9adb620c5346eba2f70f71bd50ed4f85a5e2844ac130e112165ce35c06R118-R119

clokep commented on tests/msc2403_test.go:
```diff
@@ -0,0 +1,241 @@
+		inRoomUser.MustDo(
+			t,
+			"POST",
+			[]string{"_matrix", "client", "r0", "rooms", roomID, "invite"},
```
Is this a copy and paste error? Shouldn't the user be being banned, not invited?

anoadragon453 replied:
Err, yes. This is why it's handy to have different error codes per situation :(

clokep commented on tests/msc2403_test.go:
```diff
@@ -0,0 +1,241 @@
+		knockOnRoom(t, knockingUser, roomID, "Pleeease let me in?", []string{"hs1"})
+	})
+
+	t.Run("A user in the room can accept a knock", func(t *testing.T) {
```
I think a knock after this isn't allowed and should 403?

anoadragon453 replied:
True! I'll stick in a test for that, thanks!

anoadragon453 commented:
@Kegsay heads up I'm looking for a review from your side as well just to take a look over whether these tests were written in a way you think works with Complement :)

clokep commented on tests/msc2403_test.go:
```diff
@@ -0,0 +1,281 @@
+		knockOnRoomWithStatus(t, knockingUser, roomID, reason, []string{"hs1"}, 403)
+	})
+
+	t.Run("A user that is banned from a room cannot knock on it", func(t *testing.T) {
```
Is this really testing what we think it is or is it testing that the user is already in the room? (Or does banning also kick the user?)

anoadragon453 replied:
Banning will kick a user, yes. It updates their membership to `ban`, removing them from the room.

clokep commented on tests/msc2403_test.go:
```diff
@@ -0,0 +1,281 @@
+		RoomVersion string `json:"room_version"`
+	}{
+		"private_chat",      // Set to private in order to get an invite-only room
+		"xyz.amorgan.knock", // Room version required for knocking. TODO: Remove when knocking is in a stable room version
```
What do we do when this is standardized? Do we just update these? Do we somehow let a server specify that they support the MSC vs. the standardized version?

anoadragon453 replied:
Once the MSC is merged into a released version of the spec, Complement should be updated to support that release. Part of that update will be stabilizing knocks.

The necessary steps for stabilizing will be:

* Updating instances of `xyz.amorgan.knock` to `knock`.
* Update endpoints from `/_matrix/client/unstable/xyz.amorgan.knock/...` to `/_matrix/client/rX/...`.
* Removing the `// +build ...` line at the top so the tests are run by default (they're currently opt in).
* Updating the name of the test file to `knock_test.go` or somesuch.
* Removing the conditional on rescinding knocks over federation. Synapse will need to support that if the spec says it's possible.

It might be a good idea to put the above list into the top of the file actually.

clokep replied:
I guess my broader question is -- does complement have a way for homeservers to say what version of the spec they support? Not expecting it here necessarily, just bringing it up for conversation!

And if it does, does it make sense for a homeserver to say it supports 0.6 + msc2403 even after 0.7 is released?

anoadragon453 replied:
@Kegsay I'd be interested in your answer here in case it differs from mine.

I don't think it does, and in doing so I think Complement would need to support multiple spec versions, which I don't think it's planning to. Mostly because new spec versions often require modifying existing tests instead of making a separate set of new ones.

Selecting a range of tests is done via build tags, thus I think this choice would be made in your individual CI script rather than in a config file in the server/complement (at least currently!).

clokep replied:
Alright. I don't think we need to solve this here, was just a thought I had that!

clokep reviewed (approved):
I think the logic looks sound! I didn't closely look at the changes to the blueprint/client.

Kegsay commented on tests/msc2403_test.go:
```diff
@@ -0,0 +1,281 @@
+	}
+	b, err := json.Marshal(requestBody)
+	if err != nil {
+		t.Fatalf("knockOnRoomWithStatus failed to marshal JSON body: %s", err)
```
You can use `must.NotError(t, msg, err)`

Kegsay commented on tests/msc2403_test.go:
```diff
@@ -0,0 +1,281 @@
+		inRoomUser.MustDo(
+			t,
+			"PUT",
+			[]string{"_matrix", "client", "r0", "rooms", roomID, "state", "m.room.join_rules", ""},
```
Don't do this, use `CSAPI.SendEventSynced(t, roomID, b.Event)` to send state events into the room. This will wait for the response to come down `/sync` which is critical to reduce racey tests.

Kegsay commented on tests/msc2403_test.go:
```diff
@@ -0,0 +1,281 @@
+					return false
+				}
+				must.EqualStr(t, ev.Get("content").Get("reason").Str, testKnockReason, "incorrect reason for knock")
+				must.EqualStr(t, ev.Get("content").Get("membership").Str, "xyz.amorgan.knock", "incorrect membership for knocking user")
```
I would make `const constKnock = "xyz.amorgan.knock"` and use that, which will also help for reducing the number of changes when it lands.

Kegsay commented on tests/msc2403_test.go:
```diff
@@ -0,0 +1,281 @@
+			knockingUser.MustDo(
+				t,
+				"POST",
+				[]string{"_matrix", "client", "r0", "rooms", roomID, "leave"},
```
You should probably wait for the leave to come down sync.

Kegsay commented on tests/msc2403_test.go:
```diff
@@ -0,0 +1,281 @@
+		inRoomUser.MustDo(
+			t,
+			"POST",
+			[]string{"_matrix", "client", "r0", "rooms", roomID, "kick"},
```
Wait for it to come down sync before knocking again.

Kegsay commented on tests/msc2403_test.go:
```diff
@@ -0,0 +1,281 @@
+		inRoomUser.MustDo(
+			t,
+			"POST",
+			[]string{"_matrix", "client", "r0", "rooms", roomID, "ban"},
```
Wait for it to come down sync.

Kegsay reviewed (changes requested):
Few things but broadly LGTM

anoadragon453 commented:
@Kegsay quite a few chunky changes since your last review, all addressing your feedback. Details are in commit messages, but overall it was mostly a lot of code for syncing, part of which required some slight changes to client.go.

Kegsay reviewed (approved):
We'll probably want to factor out the leave synced stuff so we don't need to hit initial sync so much but for now this will do. LGTM!
